aten/src/ATen/BatchingRegistrations.cpp

```diff
@@ -238,6 +258,15 @@ Tensor diagonal_batching_rule(const Tensor& self, int64_t offset, int64_t dim1,
   return self_physical.newLogicalFromPhysical(result);
 }
 
+Tensor diagonal_backward_batching_rule(const Tensor& grad, IntArrayRef input_sizes, int64_t offset, int64_t dim1, int64_t dim2) {
+  auto grad_physical = MultiBatchVmapTransform::logicalToPhysical(grad);
+  auto grad_input = at::zeros(grad_physical.getPhysicalShape(input_sizes), grad.options());
```

Comment:
Do you really want to keep the grad's layout here? Or you want a contiguous layout (to make sure the .diagonal() call below won't fail)?

Reply:
Same for the other functions actually

Reply:
The options for memory format right now are all dense (contiguous memory format vs channels last memory format), so the diagonal() call will never fail. Also, if there is something wrong here, then we probably want to change https://github.com/pytorch/pytorch/blob/4bead6438a4271f154980dc242199c2f1b897bc1/torch/csrc/autograd/FunctionsManual.cpp#L1120-L1125 as well...

Reply:
Ok that sounds good.
I would guess that this is not tested though.
A simple thing like `a = torch.rand(10, 10, requires_grad=True); a.diagonal().sum().backward(); self.assertAllClose(a.grad, torch.eye(10))` could be added.
Do you think this can be added here or we should do that in a follow up PR?

Reply:
There are tests for diagonal_backward: https://github.com/pytorch/pytorch/blob/a2a81e1335f28dbd21ef517f18ff218706e70ffd/torch/testing/_internal/common_methods_invocations.py#L918-L929

Are you suggesting that we add a test for the case that `grad` is an expanded tensor? So something like:
```
a = torch.rand(10, 10, requires_grad=True)
constant = 3.14
v = torch.tensor([constant]).expand(10)
a_grad = torch.autograd.grad(a.diagonal(), a, v)
self.assertAllClose(a_grad, torch.eye(10) * constant)
```

Reply:
Yes something like that. That will not cover all the cases but would cover at least this one. You don't think that would be useful?

Reply:
I agree that that would be useful, I'll add it in this PR :)

Reply:
Hmm, but in general, it might be nice for gradcheck to pass in an expanded or non-contiguous `v` tensor to autograd.grad

Reply:
It's not really possible in general because we pass a very specific value (vector of zeros with a single one) and that cannot be done with an expanded Tensor.
But the gradgradcheck, you can generate non-contiguous grad_outputs. So this is tested for the functions where we do gradgradcheck.

Reply:
Updated with tests for torch.diagonal, torch.select, and torch.slice in the case where `v` is an expanded tensor